### ISSUE TITLE
Pennant Exercising a DMA Bug

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#234]
Pennant (without control replication) is exercising a bug in the new DMA system in the master branch.  Below is a command line that runs correctly in the 'olddma' branch, but results in bad data (a pointer in this case) being read from a field in the 'master' branch.  Notice that command line uses '-lg:inorder' to ensure that the runtime is doing things in strict program order.  Also notice the '-fbounds-checks' means this code takes a while to run so be patient.

Build regent with: ./install.py --debug --gasnet

LAUNCHER='mpirun -H n0002,n0003 -np 4 --map-by numa:pe=6 -x LD_LIBRARY_PATH -x INCLUDE_PATH -x TERRA_PATH --oversubscribe' ./regent.py examples/pennant_fast.rg examples/pennant.tests/leblanc/leblanc.pnt -ll:cpu 5 -npieces 20 -numpcx 1 -numpcy 20  -par_init 1 -seq_init 0 -fbounds-checks 1 -lg:inorder

When the code fails in the 'master' branch you will see the following error:
Errors reported during runtime.
examples/pennant_fast.rg:122: pointer ptr(zone(), $rz) is out-of-bounds

@magnatelee first reported this bug so ask him to test it when there is a fix.

**Comment by lightsighter** [comment]
By commenting out the application assertions I was able to validate the runtime execution with Legion Spy, so I'm now confident that the issue is in the DMA system.

I've also noticed other weird failure modes where it is not just the pointer data that is invalid, suggesting that copies of any field are subject to having their data corrupted.

**Comment by lightsighter** [comment]
I'm about to assign the bug back to Sean and Zhihao, but before I do here is a summary of what we learned and why we now know the bug is in the DMA system (both new AND old have the same bug).  We've simplified the test case to a single node and it can be reproduced with the following command line:  

./regent.py examples/pennant_fast.rg examples/pennant.tests/leblanc/leblanc.pnt -ll:cpu 2 -npieces 2 -numpcx 1 -numpcy 2 -par_init 1 -seq_init 0 -lg:inorder -fbounds-checks 1

1. We know this is the commit that causes the problem: 390235e  It changes the sizes of the physical instances being used.  
2. The same behavior can be observed with both the master branch and the olddma branch, so it is nothing related to the new DMA system specifically.
3. The application is passing with -fbounds-checks 1 so there is no corruption of the data by the application itself.
4. Legion Spy validates this execution as being correct by the Legion runtime and @TWarszawski and I have hand-checked about 75% of the graph ourselves.
5. The first task to fail is the calc_volumes_full task during the start-up and it fails the 'sv > 0.0' assertion.
6. The first cell to fail is cell at location 1768 in the first (and only span).
7. The reason it fails is that (deterministically) different data is being read from the fields 'zv.x' and 'zv.y' during the computation of the 'sa' temporary.  Specifically we see the following values for the computation of 'sa' in the 'good' and 'bad' cases:

sa good: 0.0025 = 0.5 * cross(<0.3,4.4> - <0.2,4.4>,<0.25,4.45> - <0.2,4.4>)
sa bad: -0.05375 = 0.5 * cross(<0.3,4.4> - <0.2,4.4>,<0.175,3.325> - <0.2,4.4>)

8. If we examine the event graph for a bad run we see that the data for this field is placed in a concrete instance and never moved.  No copies to or from this instance are issued by the runtime before the failure, which is correct.  The proper event dependences ensure that all tasks associated with this field are running in the correct order which is consistent with Legion Spy validation.
9. I traced the bad value back to 'calc_centers_full' where I determined that the difference comes from the computation of 'zx'.  On the third iteration for zone 442 the good and the bad differ while still having the same pointers for p1.

Good:
p1_px= <0.2,4.4> p1=497
p1_px= <0.3,4.4> p1=498
p1_px= <0.3,4.5> p1=3
p1_px= <0.2,4.5> p1=2
znump 4
zx 442 is <0.25,4.45> nside=4

Bad:
p1_px=<0.2,4.4> p1=497
p1_px=<0.3,4.4> p1=498
p1_px=<0,0> p1=3
p1_px=<0.2,4.5> p1=2
znump 4
zx 442 is <0.175,3.325> nside=4

10. The p1.px field is initialized by 'intiialize_topology' so I went and printed out the values that are computed for the 'px' field for location p1=3 for both the good and bad versions:

good: p1=3 = <0.3,4.5> Shared Bottom
bad: p1=3 = <0.3,4.5> Shared Bottom

Good news! They're the same, that means that something is going wrong between the tasks.

11. If you look at the attached event graph you will see that there is a copy needed between the instances used by 'initialize_topology' and 'calc_centers_full'.  Legion correctly issues this copy (Realm copy 19).  Legion Spy indicates that for this particular execution both index spaces 10 and 12 are exactly the same, so their intersection is the same (the copy is an intersection of index spaces 10 and 12) so all of the data should be moved.  

Reducing top-level index space shapes...
Done
Space Index Space 10 has 3 points
  Points: 3 7 6
Space Index Space 12 has 3 points
  Points: 3 7 6

However, clearly, p1=3 is not being moved properly, and that is a DMA system bug.
[bad.pdf](https://github.com/StanfordLegion/legion/files/877706/bad.pdf)

**Comment by lightsighter** [comment]
Note that Legion is doing the intersection computation using Realm primitives so you might want to check this code too:

https://github.com/StanfordLegion/legion/blob/master/runtime/legion/region_tree.cc#L5177-L5188

**Comment by streichler** [comment]
I've managed to reproduce the error, and am digging into it now.

**Comment by streichler** [comment]
I'm still sorting through some stuff, but there is only a single copy that occurs before calc_centers_full reads the bad value, and it appears to be copying an indexspace that is the non-empty intersection of two index spaces into an instance that was created from an empty index space.  Realm does not verify that the domain used for a copy is a subset of the domains that exist in source and destination instances (because it is too expensive), but if there were such a check, it'd be complaining loudly here.

Some snippets from the log:
```
[0 - b0910000] {2}{region}: subregion 500000000000000a (of 5000000000000007) restricted to [0,10]
[0 - b0910000] {2}{meta}: index space created: id=500000000000000a parent=5000000000000007 (num_elmts=1001)
[0 - b0910000] {2}{region}: subregion 500000000000000c (of 5000000000000007) restricted to [0,10]
[0 - b0910000] {2}{meta}: index space created: id=500000000000000c parent=5000000000000007 (num_elmts=1001)
[0 - b0910000] {2}{region}: subregion 500000000000000d (of 5000000000000007) restricted to [-1,-1]
[0 - b0910000] {2}{meta}: index space created: id=500000000000000d parent=5000000000000007 (num_elmts=1001)

[0 - b0d16000] {2}{inst}: local instance 6000000000000007 created in memory 1e00000000000000 at offset 536775424+72 (redop=0 list_size=-1 parent_inst=0 block_size=4)
[0 - b0d16000] {2}{meta}: instance created: region=500000000000000d memory=1e00000000000000 id=6000000000000007 bytes=72

[0 - b050a000] {3}{index_spaces}: creating intersection: 500000000000000a & 500000000000000c
[0 - b050a000] {2}{region}: subregion 5000000000000026 (of 0) restricted to [0,10]
[0 - b050a000] {2}{meta}: index space created: id=5000000000000026 num_elmts=64
[0 - b050a000] {1}{dma}: copy: 1 distinct src/dst mem pairs, is=5000000000000026
[0 - b050a000] {2}{dma}: dma request 0x7fb3441ecc40 created - is=5000000000000026 before=0 after=8000000002c00008
[0 - b050a000] {2}{dma}: dma request 0x7fb3441ecc40 field: 6000000000000002[0]->6000000000000007[0] size=8 serdez=0
[0 - b050a000] {2}{dma}: dma request 0x7fb3441ecc40 field: 6000000000000002[8]->6000000000000007[8] size=8 serdez=0
```

**Comment by lightsighter** [comment]
@streichler Are you sure you got the right point for calc_centers_full (there are two of them) and the right copy?  I think the copy we're interested in here should be from 6000000000000002 to 600000000000000b and not 6000000000000007.  If you're running with -lg:inorder all the mapping decisions should be deterministic and therefore the instance creation names too.  At least they were for me when I was debugging.  Instance 600000000000000b should be an instance of index space 10 which is non-empty.

**Comment by streichler** [comment]
This was the _only_ copy to have completed before the first calc_centers_full executed.  (I modified the test to assert on bad data in there, so the other instance of calc_centers_full never executes either.)  The only other copy that has even been requested at that point is dependent on the first calc_centers_full instance's completion, and appears to target the same empty instance.

**Comment by streichler** [comment]
Reassigning this to @lightsighter .  There's definitely something wrong in the instance creation.  Here's a reformatted excerpt of `legion_spy.py -i` on my trimmed down test:
```
Instance 0x6000000000000000  Region (4,1,1)
Instance 0x6000000000000001  Region (8,2,2)
Instance 0x6000000000000002  Region (12,2,2)
Instance 0x6000000000000003  Region (10,2,2)
Instance 0x6000000000000004  Region (14,3,3)
Instance 0x6000000000000005  Region (5,1,1)
Instance 0x6000000000000006  Region (9,2,2)
Instance 0x6000000000000007  rp_all (2,2,2)
Instance 0x6000000000000008  Region (11,2,2)
Instance 0x6000000000000009  Region (15,3,3)
Instance 0x600000000000000a  Region (18,4,4)
Instance 0x600000000000000b  Region (22,5,5)
Instance 0x600000000000000c  Region (26,6,6)
Instance 0x600000000000000d  Region (30,7,7)
Instance 0x600000000000000e  Region (19,4,4)
Instance 0x600000000000000f  Region (23,5,5)
Instance 0x6000000000000010  Region (27,6,6)
Instance 0x6000000000000011  Region (31,7,7)
Instance 0x6000000000000012  Region (4,1,1)
Instance 0x6000000000000013  Region (5,1,1)
Instance 0x6000000000000014  Region (14,3,3)
```

And here's the Realm logging of instance creations:
```
[0 - b0284000] {2}{meta}: instance created: region=5000000000000004 memory=1e00000000000000 id=6000000000000000 bytes=452
[0 - b0284000] {2}{meta}: instance created: region=5000000000000008 memory=1e00000000000000 id=6000000000000001 bytes=8928
[0 - b0284000] {2}{meta}: instance created: region=500000000000000c memory=1e00000000000000 id=6000000000000002 bytes=216
[0 - b0284000] {2}{meta}: instance created: region=500000000000000a memory=1e00000000000000 id=6000000000000003 bytes=192
[0 - b0284000] {2}{meta}: instance created: region=500000000000000e memory=1e00000000000000 id=6000000000000004 bytes=75600
[0 - b0d16000] {2}{meta}: instance created: region=5000000000000005 memory=1e00000000000000 id=6000000000000005 bytes=452
[0 - b0d16000] {2}{meta}: instance created: region=5000000000000009 memory=1e00000000000000 id=6000000000000006 bytes=8928
[0 - b0d16000] {2}{meta}: instance created: region=500000000000000d memory=1e00000000000000 id=6000000000000007 bytes=72
[0 - b0d16000] {2}{meta}: instance created: region=500000000000000b memory=1e00000000000000 id=6000000000000008 bytes=192
[0 - b0d16000] {2}{meta}: instance created: region=500000000000000f memory=1e00000000000000 id=6000000000000009 bytes=75600
[0 - b111c000] {2}{meta}: instance created: region=5000000000000011 memory=1e00000000000000 id=600000000000000a bytes=68
[0 - b111c000] {2}{meta}: instance created: region=5000000000000014 memory=1e00000000000000 id=600000000000000b bytes=68
[0 - b111c000] {2}{meta}: instance created: region=5000000000000017 memory=1e00000000000000 id=600000000000000c bytes=68
[0 - b111c000] {2}{meta}: instance created: region=500000000000001a memory=1e00000000000000 id=600000000000000d bytes=68
[0 - b0284000] {2}{meta}: instance created: region=5000000000000012 memory=1e00000000000000 id=600000000000000e bytes=68
[0 - b0284000] {2}{meta}: instance created: region=5000000000000015 memory=1e00000000000000 id=600000000000000f bytes=68
[0 - b0284000] {2}{meta}: instance created: region=5000000000000018 memory=1e00000000000000 id=6000000000000010 bytes=68
[0 - b0284000] {2}{meta}: instance created: region=500000000000001b memory=1e00000000000000 id=6000000000000011 bytes=68
[0 - b0284000] {2}{meta}: instance created: region=5000000000000004 memory=1e00000000000000 id=6000000000000012 bytes=14464
[0 - b131f000] {2}{meta}: instance created: region=5000000000000005 memory=1e00000000000000 id=6000000000000013 bytes=14464
[0 - b0284000] {2}{meta}: instance created: region=500000000000000e memory=1e00000000000000 id=6000000000000014 bytes=28800
```

It looks like the Legion and Realm index space numbers deviate after a while, but all of the first ones match up EXCEPT for instance `6000000000000007`, which happens to be the target of the copy, and for which that copy would make a lot more sense if the instance had been constructed from `rp_all` instead of `(13,2,2)` (which looks like it's rp_shared[1] from the index space tree).

**Comment by lightsighter** [comment]
I believe that this bug is now fixed with da5e088. Assigning back to @magnatelee to confirm.

**Comment by magnatelee** [comment]
The bug seems fixed. Close this issue.
